### PR TITLE
[FSDP] Deduplicate `_orig_size` and `_unsharded_size`

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #81243 [FSDP] Move `_post_backward_called` to `_init_param_attributes`
* #80185 [FSDP] Clean up `_lazy_init()`
* #80087 [Easy][FSDP] Add `zero_grad()` to unit test train loop
* #80002 [FSDP] Remove `self.numel_padded_per_param` (unused)
* #80000 [FSDP] Move tensor sharding logic to `FlatParamHandle`
* **#79984 [FSDP] Deduplicate `_orig_size` and `_unsharded_size`**
* #79652 [FSDP] Introduce `FlatParamHandle`

This removes the `_orig_size` attribute that is initialized in `fully_sharded_data_parallel.py` since it represents the same quantity as `_unsharded_size` in `flat_param.py`. Since the quantity is not sharding dependent, we keep its initialization in `FlatParameter.init_metadata()` instead of in `FullyShardedDataParallel._shard_parameters()`.

Differential Revision: [D37726062](https://our.internmc.facebook.com/intern/diff/D37726062)

### CODE REVIEW

File: torch/distributed/fsdp/_optim_utils.py [hunk 1]
```diff
@@ -160,9 +160,7 @@ def _communicate_optim_state(
                 tensor_buffer = value.new_zeros(*buffer_size)
             dist._all_gather_base(tensor_buffer, value, group=group)
             if to_save:
-                assert hasattr(flat_param, "_orig_size"), \
```
Comment (awgu) [inline torch/distributed/fsdp/_optim_utils.py:163]: The `assert` is not useful. Every `FlatParameter` should call `init_metadata()`, so `_unsharded_size` is guaranteed to be an attribute.

**Review by rohan-varma — APPROVED** [review]

**Comment by rohan-varma** [comment]
LGTM

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37726062).

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37726062).

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37726062).

**Pull request closed without merging**